src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py

```diff
@@ -1077,26 +1172,69 @@ def write_mets(tree, filename):
 
     baseDirectoryPath = os.path.join(baseDirectoryPath, '')
     objectsDirectoryPath = os.path.join(baseDirectoryPath, 'objects')
+    objectsMetadataDirectoryPath = os.path.join(objectsDirectoryPath, 'metadata')
 
     # Delete empty directories, see #8427
     for root, dirs, files in os.walk(baseDirectoryPath, topdown=False):
         try:
             os.rmdir(root)
-            print("Deleted empty directory", root)
+            # If we just deleted an empty subdirectory of objects/ (but not of
+            # objects/metadata/), then re-create it with a .keep file in it.
+            # (This deletion and re-creation seems more efficient than calling
+            # ``os.listdir`` on every directory to see if it is empty.)
+            if (root.startswith(objectsDirectoryPath) and
+                    root != objectsDirectoryPath and
+                    not root.startswith(objectsMetadataDirectoryPath)):
+                os.makedirs(root)
+                fpath = os.path.join(root, '.keep')
+                with open(fpath, 'a'):
+                    os.utime(fpath, None)
+            else:
+                print("Deleted empty directory", root)
         except OSError:
             pass
 
-    structMap = etree.Element(ns.metsBNS + "structMap", TYPE='physical', ID='structMap_1', LABEL="Archivematica default")
-    structMapDiv = etree.SubElement(structMap, ns.metsBNS + 'div', TYPE="Directory", LABEL=os.path.basename(baseDirectoryPath.rstrip('/')))
-    structMapDivObjects = createFileSec(objectsDirectoryPath, structMapDiv, baseDirectoryPath, baseDirectoryPathString, fileGroupIdentifier, fileGroupType, includeAmdSec)
+    # Fetch any ``Directory`` objects in the database that are contained within
+    # this SIP and return them as a dict from relative paths to UUIDs. (See
+    # createSIPfromTransferObjects.py for the association of ``Directory``
+    # objects to a ``SIP``.
+    directories = {
+        d.currentlocation: d for d in
+        Directory.objects.filter(sip_id=fileGroupIdentifier).all()}
+
+    structMap = etree.Element(
+        ns.metsBNS + "structMap", TYPE='physical', ID='structMap_1',
+        LABEL="Archivematica default")
+    sip_dir_name = os.path.basename(baseDirectoryPath.rstrip('/'))
+    structMapDiv = etree.SubElement(
+        structMap, ns.metsBNS + 'div', TYPE="Directory",
+        LABEL=sip_dir_name)
+
+    # Get the <dmdSec> for the entire AIP; it is associated to the root
+    # <mets:div> in the physical structMap.
+    sip_mdl = SIP.objects.filter(uuid=fileGroupIdentifier).first()
+    if sip_mdl:
+        aipDmdSec = getDirDmdSec(sip_mdl, sip_dir_name)
+        globalDmdSecCounter += 1
+        dmdSecs.append(aipDmdSec)
+        aip_dmd_id = "dmdSec_" + globalDmdSecCounter.__str__()
```

Comment:
True. As with the `locals()` issue, this was a result of being consistent with the rest of the codebase, even if it is non-idiomatic, just like when I use camelCase to match the style of legacy code, etc. But I'll change it.